File: android/app/src/main/java/net/bradball/teetimecaddie/android/ui/common/forms/TtcTextField.kt
package net.bradball.teetimecaddie.android.ui.common.forms

import android.content.res.Configuration
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import net.bradball.teetimecaddie.android.theme.MyApplicationTheme
import net.bradball.teetimecaddie.android.ui.common.icons.TtcIcons

/**
 * A single-line text field from the Fairway Morning design system.
 *
 * Renders the Material 3 *filled* field: a `surfaceContainerHighest` fill with top-rounded corners
 * and a bottom indicator that tints to `primary` on focus (or `error` when [error] is set), with a
 * floating [label]. [leadingIcon], [placeholder], and [hint] are all optional. When [error] is
 * non-null the field enters its error state and shows the message in place of [hint].
 *
 * For a full-width field, pass `Modifier.fillMaxWidth()`.
 *
 * @param value The current text.
 * @param onValueChange Invoked with the new text on each edit.
 * @param label The floating field label (always shown).
 * @param modifier Modifier for the field.
 * @param leadingIcon Optional leading [TtcIcons] icon, tinted with the label color.
 * @param placeholder Optional placeholder shown once the label floats.
 * @param hint Optional supporting/helper text shown beneath the field (hidden when [error] is set).
 * @param error Optional error message. When non-null the field renders in its error state.
 * @param enabled Whether the field is enabled.
 * @param secure When true, the input is masked (e.g. for a password). See [TtcPasswordField], which
 *   wraps this field to add a reveal toggle.
 * @param trailingIcon Optional trailing [TtcIcons] icon rendered as a tappable [IconButton].
 * @param onTrailingClick Invoked when the trailing icon is tapped (ignored when [trailingIcon] is null).
 * @param trailingContentDescription Accessibility description for the trailing icon.
 * @param keyboardOptions Software-keyboard configuration (type, capitalization, IME action, …).
 */
@Composable
fun TtcTextField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    leadingIcon: TtcIcons? = null,
    placeholder: String? = null,
    hint: String? = null,
    error: String? = null,
    enabled: Boolean = true,
    secure: Boolean = false,
    trailingIcon: TtcIcons? = null,
    onTrailingClick: (() -> Unit)? = null,
    trailingContentDescription: String? = null,
    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
) {
    TextField(
        value = value,
        onValueChange = onValueChange,
        modifier = modifier,
        enabled = enabled,
        label = { Text(label) },
        placeholder = placeholder?.let { { Text(it) } },
        leadingIcon = leadingIcon?.let {
            {
                Icon(
                    painter = it.painter,
                    contentDescription = null,
                    modifier = Modifier.size(TtcTextFieldDefaults.IconSize),
                )
            }
        },
        trailingIcon = trailingIcon?.let {
            {
                IconButton(onClick = { onTrailingClick?.invoke() }) {
                    Icon(
                        painter = it.painter,
                        contentDescription = trailingContentDescription,
                        modifier = Modifier.size(TtcTextFieldDefaults.IconSize),
                    )
                }
            }
        },
        supportingText = supportingText(error = error, hint = hint),
        isError = error != null,
        visualTransformation = if (secure) PasswordVisualTransformation() else VisualTransformation.None,
        singleLine = true,
        shape = TtcTextFieldDefaults.Shape,
        colors = TtcTextFieldDefaults.colors(),
        keyboardOptions = keyboardOptions,
    )
}

/**
 * Builds the supporting-text slot: the [error] message (when set) otherwise the [hint], or null when
 * neither is present. The error message inherits the error color from [TtcTextFieldDefaults.colors].
 *
 * Note: the Fairway Morning error row includes a small leading error glyph. `TtcIcons` has no error
 * icon yet (and the full Material Icons library is intentionally not imported), so the glyph is
 * omitted for now — see the plan's follow-ups.
 */
private fun supportingText(error: String?, hint: String?): (@Composable () -> Unit)? {
    val text = error ?: hint ?: return null
    return { Text(text) }
}

@Preview(name = "Light", showBackground = true)
@Preview(name = "Dark", showBackground = true, uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun TtcTextFieldPreview() {
    MyApplicationTheme {
        Surface(color = MaterialTheme.colorScheme.surface) {
            Column(
                modifier = Modifier.padding(16.dp).fillMaxWidth(),
                verticalArrangement = Arrangement.spacedBy(14.dp),
            ) {
                TtcTextField(
                    value = "",
                    onValueChange = {},
                    label = "Email",
                    leadingIcon = TtcIcons.CALENDAR,
                    placeholder = "you@golf.app",
                    modifier = Modifier.fillMaxWidth(),
                )
                TtcTextField(
                    value = "Dana Park",
                    onValueChange = {},
                    label = "Full name",
                    modifier = Modifier.fillMaxWidth(),
                )
                TtcTextField(
                    value = "e.g. Pebble Beach",
                    onValueChange = {},
                    label = "Course",
                    hint = "We match case-insensitively against existing courses.",
                    modifier = Modifier.fillMaxWidth(),
                )
            }
        }
    }
}

@Preview(name = "States - Light", showBackground = true)
@Preview(name = "States - Dark", showBackground = true, uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun TtcTextFieldStatesPreview() {
    MyApplicationTheme {
        Surface(color = MaterialTheme.colorScheme.surface) {
            Column(
                modifier = Modifier.padding(16.dp).fillMaxWidth(),
                verticalArrangement = Arrangement.spacedBy(14.dp),
            ) {
                TtcTextField(
                    value = "(555) 010-1010",
                    onValueChange = {},
                    label = "Mobile number",
                    error = "Already linked to another account",
                    modifier = Modifier.fillMaxWidth(),
                )
                TtcTextField(
                    value = "Disabled",
                    onValueChange = {},
                    label = "Course",
                    enabled = false,
                    modifier = Modifier.fillMaxWidth(),
                )
            }
        }
    }
}

File: ios/TeeTimeCaddie/TeeTimeCaddie/Source/ui/common/TextFields/TtcTextField.swift
import SwiftUI
import ThemeUI

/// A single-line text field from the Fairway Morning design system.
///
/// iOS renders the custom "filled-row" card (matching the Create Game screen): a `surfaceContainer`
/// rounded card with a static [label] above the input, an optional [leadingIcon], and a 2pt border
/// that tints to `primary` on focus (or `error` when [error] is set). [placeholder], [hint], and
/// [leadingIcon] are all optional. When [error] is non-null the field enters its error state and
/// shows the message (in place of [hint]) beneath the card.
///
/// The card chrome (label, border, support row) is shared with ``TtcPasswordField`` via
/// ``TtcTextFieldContainer``; this field injects a single plain `TextField` as its control. Disabled
/// state follows the standard SwiftUI idiom — apply `.disabled(true)` at the call site.
struct TtcTextField: View {
    @FocusState private var isFocused: Bool

    private let label: String
    @Binding private var text: String
    private let leadingIcon: ImageResource?
    private let placeholder: String?
    private let hint: String?
    private let error: String?
    private let keyboardType: UIKeyboardType
    private let textContentType: UITextContentType?
    private let autocapitalization: TextInputAutocapitalization
    private let autocorrectionDisabled: Bool

    init(
        _ label: String,
        text: Binding<String>,
        leadingIcon: ImageResource? = nil,
        placeholder: String? = nil,
        hint: String? = nil,
        error: String? = nil,
        keyboardType: UIKeyboardType = .default,
        textContentType: UITextContentType? = nil,
        autocapitalization: TextInputAutocapitalization = .never,
        autocorrectionDisabled: Bool = true
    ) {
        self.label = label
        self._text = text
        self.leadingIcon = leadingIcon
        self.placeholder = placeholder
        self.hint = hint
        self.error = error
        self.keyboardType = keyboardType
        self.textContentType = textContentType
        self.autocapitalization = autocapitalization
        self.autocorrectionDisabled = autocorrectionDisabled
    }

    var body: some View {
        TtcTextFieldContainer(
            label: label,
            leadingIcon: leadingIcon,
            hint: hint,
            error: error,
            isFocused: isFocused,
            onActivate: { isFocused = true }
        ) {
            TextField(placeholder ?? "", text: $text)
                .modifier(
                    TtcTextFieldInputStyle(
                        isError: error?.isEmpty == false,
                        keyboardType: keyboardType,
                        textContentType: textContentType,
                        autocapitalization: autocapitalization,
                        autocorrectionDisabled: autocorrectionDisabled
                    )
                )
                .focused($isFocused)
        }
    }
}

// MARK: - Previews

#Preview("Light") {
    TeeTimeCaddieTheme {
        TtcTextFieldPreviews()
    }
}

#Preview("Dark") {
    TeeTimeCaddieTheme {
        TtcTextFieldPreviews()
    }
    .preferredColorScheme(.dark)
}

fileprivate struct TtcTextFieldPreviews: View {
    @State private var email = ""
    @State private var name = "Dana Park"
    @State private var course = "e.g. Pebble Beach"
    @State private var phone = "(555) 010-1010"

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            TtcTextField(
                "Email",
                text: $email,
                leadingIcon: .Icons.calendar,
                placeholder: "you@golf.app",
                keyboardType: .emailAddress
            )
            TtcTextField("Full name", text: $name, autocapitalization: .words)
            TtcTextField(
                "Course",
                text: $course,
                hint: "We match case-insensitively against existing courses."
            )
            TtcTextField(
                "Mobile number",
                text: $phone,
                error: "Already linked to another account",
                keyboardType: .phonePad
            )
            TtcTextField("Disabled", text: .constant("Disabled"))
                .disabled(true)
        }
        .padding()
    }
}
